File: Server/Server/Network/PacketHandler/LdpPacketHandler.cs
using Server.Network.PacketSender;
using Server.Network.PacketTypes;
using Server.WindowsUtils;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Server.Network.PacketHandler
{
    abstract class LdpPacketHandler : ILdpObservable, IDisposable
    {
        private static object MUTEX = new object();
        protected abstract void Handle();

        private List<ILdpPacketHandler> listeners;
        public LdpPacketHandler()
        {
            listeners = new List<ILdpPacketHandler>();
        }

        #region Observer pattern impl
        public void AddListener(ILdpPacketHandler listener)
        {
            lock (MUTEX)
            {
                if (!listeners.Contains(listener))
                    listeners.Add(listener);
            }
        }

        public void RemoveListener(ILdpPacketHandler listener)
        {
            lock (MUTEX)
            {
                if (listeners.Contains(listener))
                {
                    listeners.Remove(listener);
                    LdpLog.Info(String.Format("Listener removed: {0}.", listener.GetType()));
                }
            }
        }

        public void RemoveListeners()
        {
            lock (MUTEX)
            {
                if (listeners != null)
                {
                    listeners.Clear();
                    listeners = null;
                }
            }
        }

        public void NotifyToAllListeners(LdpPacket packet)
        {
            if (listeners != null && listeners.Count != 0)
            {
                foreach (var listener in listeners.ToList())
                {
                    ILdpPacketHandler ipl = (ILdpPacketHandler)listener;
                    ipl.Handle(packet);
                }
            }
        }
        #endregion

        public void Dispose()
        {
            RemoveListeners();
        }
    }
}

File: AndroidClient/LdpAndroidClient/app/src/main/java/com/ldp/androidclient/network/packet_handler/LdpPacketHandler.java
package com.ldp.androidclient.network.packet_handler;


import com.ldp.androidclient.protocol.*;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.concurrent.CopyOnWriteArrayList;

abstract public class LdpPacketHandler implements ILdpObservable {
    private static final Object MUTEX = new Object();

    protected abstract void handle();

    private CopyOnWriteArrayList<ILdpPacketHandler> listeners;

    public LdpPacketHandler() {
        listeners = new CopyOnWriteArrayList<ILdpPacketHandler>();
    }

    @Override
    public void addListener(ILdpPacketHandler listener) {
        synchronized (MUTEX) {
            if (!listeners.contains(listener))
                listeners.add(listener);
        }
    }

    @Override
    public void notifyToAllListeners(LdpProtocol.LdpPacket packet) {
        if (listeners != null && listeners.size() != 0) {
            Iterator<ILdpPacketHandler> iterator = listeners.iterator();
            while (iterator.hasNext()) {
                ILdpPacketHandler lis = iterator.next();
                lis.handle(packet);
            }
        }
    }

    @Override
    public void removeListener(ILdpPacketHandler listener) {
        synchronized (MUTEX) {
            if (listeners.contains(listener))
                listeners.remove(listener);
        }
    }

    @Override
    public void removeListeners() {
        synchronized (MUTEX) {
            if (listeners != null) {
                listeners.clear();
                listeners = null;
            }
        }
    }
}
